tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,289 @@
+            var version = diagnosticSourceAssemblyName.Version;
+
+            // Check version: First version where the Activity objects has traceId and spanId
+            if (version >= new Version(4, 0, 4))
+            {
+                if (activityType is null)
+                {
+                    Log.Error("The Activity type cannot be found.");
+                    return;
+                }
+
+                CreateCurrentActivityDelegates(activityType);
+                ChangeActivityDefaultFormat(activityType);
+
+                if (version.Major >= 5)
+                {
+                    // if Version >= 5 loaded (Uses ActivityListener implementation)
+                    CreateActivityListenerInstance(activityType);
+                }
+                else
+                {
+                    // if Version is 4.0.4 or greater (Uses DiagnosticListener implementation / Nuget version 4.6.0)
+                    CreateDiagnosticSourceListenerInstance(diagnosticListenerType);
+                }
+
+                return;
+            }
+
+            Log.Information("An activity listener was found but version {version} is not supported.", version?.ToString() ?? "(null)");
+
+            static void CreateCurrentActivityDelegates(Type activityType)
+            {
+                // Create Activity.Current delegate.
+                var activityCurrentMethodInfo = activityType.GetProperty("Current", BindingFlags.Public | BindingFlags.Static)?.GetMethod;
+                if (activityCurrentMethodInfo is null)
+                {
+                    throw new NullReferenceException("Activity.Current property cannot be found in the Activity type.");
+                }
+
+                var activityCurrentDynMethod = new DynamicMethod("ActivityCurrent", typeof(object), Type.EmptyTypes, typeof(ActivityListener).Module, true);
+                var activityCurrentDynMethodIl = activityCurrentDynMethod.GetILGenerator();
+                activityCurrentDynMethodIl.EmitCall(OpCodes.Call, activityCurrentMethodInfo, null);
+                activityCurrentDynMethodIl.Emit(OpCodes.Ret);
+                _getCurrentActivity = (Func<object>)activityCurrentDynMethod.CreateDelegate(typeof(Func<object>));
+            }
+
+            static void ChangeActivityDefaultFormat(Type activityType)
+            {
+                // We change the default ID format to W3C (so traceid and spanid gets populated)
+                if (Activator.CreateInstance(activityType, string.Empty).TryDuckCast<IActivityFormat>(out var activityFormat))
+                {
+                    activityFormat.DefaultIdFormat = ActivityIdFormat.W3C;
+                }
+            }
+        }
+
+        public static void StopListeners()
+        {
+            // If there's an activity listener instance we detach the instance and clear it.
+            if (_activityListenerInstance is { } activityListenerInstance)
```

Comment:
Would it be better to cast this as `IDisposable` and call `Dipose()`, instead of calling the internal `ActivitySource.DetachListener(ActivityListener)`?

Reply:
you're totally right, done here: https://github.com/DataDog/dd-trace-dotnet/pull/2446/commits/1b31afc9171b3c89fe4d2dd64dae77c7028e2c76